Editor code editor issue with copy paste

Not sure how it started, but during a session the editor gave an exception and after that the copy paste stopped working. It seems all other functionalities are the same.

I haven't found a workaround, but had to restart the editor.

```
...

2026-01-18 07:55:40.208 79056581 [Timer-1] INFO  editor.updater - {:line 310, :message "New version found", :sha1 "d12c3e646c36957930eb8b250d466f8b7cbc470e"}
2026-01-18 08:59:24.413 82880786 [JavaFX Application Thread] INFO  editor.defold-project - {:line 385, :message "Cached loaded save data in system cache.", :total 298, :retained 140, :unretained 158, :limit 20000}
2026-01-18 09:01:31.929 148  [JavaFX-Launcher] WARN  javafx - Unsupported JavaFX configuration: classes were loaded from 'unnamed module @4fae72d1'
2026-01-18 09:01:32.390 609  [JavaFX Application Thread] WARN  javafx.css - CSS Error parsing jar:file:/Applications/Defold_Stable.app/Contents/Resources/packages/defold-3206f699aaff89f357c9d549050b8453e080c5d2.jar!/splash.css: Unexpected at-rule [193,1]
2026-01-18 09:01:32.973 1192 [Thread-5] INFO  com.defold.libs.ResourceUnpacker - deleting old unpack dirs from /Users/mathiaswesterdahl/Library/Application Support/Defold/unpack
2026-01-18 09:01:32.979 1198 [Thread-5] INFO  com.defold.libs.ResourceUnpacker - Already unpacked for the editor version 3206f699aaff89f357c9d549050b8453e080c5d2
2026-01-18 09:01:32.981 1200 [Thread-5] INFO  com.defold.libs.ResourceUnpacker - defold.unpack.path=/Users/mathiaswesterdahl/Library/Application Support/Defold/unpack/3206f699aaff89f357c9d549050b8453e080c5d2
2026-01-18 09:01:35.946 4165 [Timer-1] INFO  editor.updater - {:line 304, :message "Checking for updates", :url "https://d.defold.com/editor2/channels/editor-alpha/update-v4.json"}
2026-01-18 09:01:36.160 4379 [Timer-1] INFO  editor.updater - {:line 311, :message "No update found"}
2026-01-18 09:01:38.306 6525 [clojure-agent-send-off-pool-11] INFO  editor.boot - {:line 80, :message "Finished loading editor namespaces.", :namespace-counter 2617}
2026-01-18 09:01:38.343 6562 [clojure-agent-send-off-pool-11] INFO  internal.low-memory - {:line 55, :message "Adding post-gc low-memory notification listener to tenured generation pool.", :threshold-megabytes 11059}
2026-01-18 09:01:39.584 7803 [clojure-agent-send-off-pool-11] INFO  editor.defold-project - {:line nil, :message "Initial resource sync completed in 45.44 ms"}
2026-01-18 09:01:39.893 8112 [clojure-agent-send-off-pool-11] INFO  editor.workspace - {:line 613, :message "Loading plugin defold-rive/editor/src/rive.clj"}
2026-01-18 09:01:40.303 8522 [clojure-agent-send-off-pool-11] INFO  editor.workspace - {:line 619, :message "Loaded plugin defold-rive/editor/src/rive.clj"}
2026-01-18 09:01:41.875 10094 [clojure-agent-send-off-pool-11] INFO  editor.defold-project - {:line 375, :message "Cached loaded save data in system cache.", :total 1182, :retained 1182, :unretained 0, :limit 20000}
2026-01-18 09:01:41.876 10095 [clojure-agent-send-off-pool-11] INFO  editor.defold-project - {:line nil, :message "Project loading completed in 1.57 s"}
2026-01-18 09:01:41.993 10212 [clojure-agent-send-off-pool-11] INFO  editor.defold-project - {:line nil, :message "Project loaded", :allocated "173.41 MB"}
2026-01-18 09:01:42.268 10487 [JavaFX Application Thread] DEBUG org.eclipse.jgit.util.FS - readpipe [git, --version],/usr/bin
2026-01-18 09:01:42.296 10515 [JavaFX Application Thread] DEBUG org.eclipse.jgit.util.FS - readpipe may return 'git version 2.50.1 (Apple Git-155)'
2026-01-18 09:01:42.296 10515 [JavaFX Application Thread] DEBUG org.eclipse.jgit.util.FS - remaining output:

2026-01-18 09:01:42.296 10515 [JavaFX Application Thread] DEBUG org.eclipse.jgit.util.FS - readpipe [git, config, --system, --edit],/usr/bin
2026-01-18 09:01:42.314 10533 [JavaFX Application Thread] DEBUG org.eclipse.jgit.util.FS - readpipe may return '/private/etc/gitconfig'
2026-01-18 09:01:42.314 10533 [JavaFX Application Thread] DEBUG org.eclipse.jgit.util.FS - remaining output:

2026-01-18 09:01:42.518 10737 [JavaFX Application Thread] INFO  util.http-server - {:line 376, :msg "Http server running", :local-url "http://localhost:62338"}
2026-01-18 09:01:42.552 10771 [clojure-agent-send-off-pool-11] INFO  editor.boot-open-project - {:line 436, :message "project loaded"}
2026-01-18 09:01:56.563 24782 [JavaFX Application Thread] INFO  editor.defold-project - {:line 375, :message "Cached compiled build targets in system cache.", :total 1371, :retained 1294, :unretained 77, :limit 20000}
2026-01-18 09:02:19.753 47972 [JavaFX Application Thread] INFO  editor.defold-project - {:line 375, :message "Cached loaded save data in system cache.", :total 1377, :retained 1293, :unretained 84, :limit 20000}
2026-01-18 09:02:20.584 48803 [JavaFX Application Thread] INFO  editor.defold-project - {:line 375, :message "Cached compiled build targets in system cache.", :total 1383, :retained 1294, :unretained 89, :limit 20000}
2026-01-18 09:02:26.461 54680 [JavaFX Application Thread] INFO  editor.defold-project - {:line 375, :message "Cached loaded save data in system cache.", :total 1365, :retained 1293, :unretained 72, :limit 20000}
2026-01-18 09:03:50.946 139165 [JavaFX Application Thread] INFO  editor.defold-project - {:line 375, :message "Cached loaded save data in system cache.", :total 1308, :retained 1220, :unretained 88, :limit 20000}
2026-01-18 09:03:50.991 139210 [JavaFX Application Thread] ERROR editor.error-reporting - {:line 115, :ex-data {:handler {:command :edit.cut, :contexts [:code-view], :active? #object[clojure.lang.AFunction$1 0x6d04d4ce "clojure.lang.AFunction$1@6d04d4ce"], :enabled? #object[clojure.lang.AFunction$1 0x5f70924f "clojure.lang.AFunction$1@5f70924f"], :run #object[clojure.lang.AFunction$1 0x79b6a44d "clojure.lang.AFunction$1@79b6a44d"]}, :command-context #editor.handler.Context{:name :code-view, :env {:view-node 360287970189639680, :selection-provider nil, :replace-bar #object[javafx.scene.layout.GridPane 0x7edbf3ff "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :editable true, :open-resource-fn #object[clojure.core$partial$fn__5933 0x75d062d0 "clojure.core$partial$fn__5933@75d062d0"], :selection-context :code-view, :goto-line-bar #object[javafx.scene.layout.GridPane 0x47f75f0e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :user-data nil, :clipboard #object[javafx.scene.input.Clipboard 0x7df480c2 "javafx.scene.input.Clipboard@7df480c2"], :prefs {:scopes {:global #object[sun.nio.fs.UnixPath 0x5ff0666b "/Users/mathiaswesterdahl/Library/Preferences/Defold/prefs.editor_settings"], :project #object[sun.nio.fs.UnixPath 0x7cb45571 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive/.editor_settings"]}, :schemas [:default #object[sun.nio.fs.UnixPath 0xa046ed5 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive"]]}, :selection nil, :_context #editor.handler.Context{:name :code-view, :env {:view-node 360287970189639680, :selection-provider nil, :replace-bar #object[javafx.scene.layout.GridPane 0x7edbf3ff "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :editable true, :open-resource-fn #object[clojure.core$partial$fn__5933 0x75d062d0 "clojure.core$partial$fn__5933@75d062d0"], :selection-context :code-view, :goto-line-bar #object[javafx.scene.layout.GridPane 0x47f75f0e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :clipboard #object[javafx.scene.input.Clipboard 0x7df480c2 "javafx.scene.input.Clipboard@7df480c2"], :prefs {:scopes {:global #object[sun.nio.fs.UnixPath 0x5ff0666b "/Users/mathiaswesterdahl/Library/Preferences/Defold/prefs.editor_settings"], :project #object[sun.nio.fs.UnixPath 0x7cb45571 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive/.editor_settings"]}, :schemas [:default #object[sun.nio.fs.UnixPath 0xa046ed5 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive"]]}, :selection nil, :find-bar #object[javafx.scene.layout.GridPane 0x41246d4e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"]}, :selection-provider nil, :dynamics {}, :adapters {}}, :find-bar #object[javafx.scene.layout.GridPane 0x41246d4e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"]}, :selection-provider nil, :dynamics {}, :adapters {}}}}
clojure.lang.ExceptionInfo: handler ':edit.cut' in context ':code-view' failed at ':enabled?' with message 'class clojure.lang.MapEntry cannot be cast to class editor.code.data.CursorRange (clojure.lang.MapEntry and editor.code.data.CursorRange are in unnamed module of loader 'app')'
	at editor.handler$invoke_fnk.invokeStatic(handler.clj:349)
	at editor.handler$enabled_QMARK_.invokeStatic(handler.clj:374)
	at editor.ui$make_menu_item.invokeStatic(ui.clj:1455)
	at editor.ui$make_menu_items$fn__30152.invoke(ui.clj:1472)
	at clojure.core$keep$fn__8687$fn__8688.invoke(core.clj:7517)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:418)
	at clojure.core$transduce.invokeStatic(core.clj:7025)
	at clojure.core$into.invokeStatic(core.clj:7042)
	at editor.ui$make_menu_items.invokeStatic(ui.clj:1470)
	at editor.ui$make_menu_items.invoke(ui.clj:1470)
	at editor.ui$make_menu_item.invokeStatic(ui.clj:1447)
	at editor.ui$make_menu_items$fn__30152.invoke(ui.clj:1472)
	at clojure.core$keep$fn__8687$fn__8688.invoke(core.clj:7517)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:418)
	at clojure.core$transduce.invokeStatic(core.clj:7025)
	at clojure.core$into.invokeStatic(core.clj:7042)
	at editor.ui$make_menu_items.invokeStatic(ui.clj:1470)
	at editor.ui$refresh_menubar_BANG_.invokeStatic(ui.clj:1718)
	at editor.ui$refresh_menus_BANG_.invokeStatic(ui.clj:2033)
	at editor.ui$refresh.invokeStatic(ui.clj:2066)
	at editor.app_view$refresh_menus_and_toolbars_BANG_.invokeStatic(app_view.clj:1967)
	at editor.app_view$make_app_view$fn__72930.invoke(app_view.clj:2103)
	at editor.ui$__GT_timer$fn__30535$fn__30536.invoke(ui.clj:2177)
	at clojure.lang.AFn.run(AFn.java:22)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$0(PlatformImpl.java:424)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
Caused by: java.lang.ClassCastException: class clojure.lang.MapEntry cannot be cast to class editor.code.data.CursorRange (clojure.lang.MapEntry and editor.code.data.CursorRange are in unnamed module of loader 'app')
	at editor.code.data$cursor_range_empty_QMARK_.invokeStatic(data.clj:205)
	at editor.code.data$cursor_range_empty_QMARK_.invoke(data.clj:204)
	at clojure.core$every_QMARK_.invokeStatic(core.clj:2698)
	at clojure.core$every_QMARK_.invoke(core.clj:2689)
	at clojure.core$comp$fn__5895.invoke(core.clj:2587)
	at editor.code.view$has_selection_QMARK_.invokeStatic(view.clj:2687)
	at editor.code.view$fn__63196$fnk63193_positional__63197.invoke(view.clj:2731)
	at editor.code.view$fn__63196$fnk63193__63208$fn__63209.invoke(view.clj:2731)
	at editor.code.view$fn__63196$fnk63193__63208.invoke(view.clj:2731)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:33)
	at clojure.lang.RestFn.invoke(RestFn.java:411)
	at editor.handler$invoke_fnk.invokeStatic(handler.clj:343)
	... 25 common frames omitted
2026-01-18 09:03:50.993 139212 [JavaFX Application Thread] ERROR editor.error-reporting - {:line 115, :ex-data {:handler {:command :code.sort-lines, :contexts [:code-view], :label #object[clojure.lang.AFunction$1 0x26583311 "clojure.lang.AFunction$1@26583311"], :options #object[clojure.lang.AFunction$1 0x6b23d70c "clojure.lang.AFunction$1@6b23d70c"], :active? #object[clojure.lang.AFunction$1 0x1e250d15 "clojure.lang.AFunction$1@1e250d15"], :enabled? #object[clojure.lang.AFunction$1 0x77055340 "clojure.lang.AFunction$1@77055340"], :run #object[clojure.lang.AFunction$1 0x1022260a "clojure.lang.AFunction$1@1022260a"]}, :command-context #editor.handler.Context{:name :code-view, :env {:view-node 360287970189639680, :selection-provider nil, :replace-bar #object[javafx.scene.layout.GridPane 0x7edbf3ff "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :editable true, :open-resource-fn #object[clojure.core$partial$fn__5933 0x75d062d0 "clojure.core$partial$fn__5933@75d062d0"], :selection-context :code-view, :goto-line-bar #object[javafx.scene.layout.GridPane 0x47f75f0e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :user-data :case-insensitive, :clipboard #object[javafx.scene.input.Clipboard 0x7df480c2 "javafx.scene.input.Clipboard@7df480c2"], :prefs {:scopes {:global #object[sun.nio.fs.UnixPath 0x5ff0666b "/Users/mathiaswesterdahl/Library/Preferences/Defold/prefs.editor_settings"], :project #object[sun.nio.fs.UnixPath 0x7cb45571 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive/.editor_settings"]}, :schemas [:default #object[sun.nio.fs.UnixPath 0xa046ed5 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive"]]}, :selection nil, :_context #editor.handler.Context{:name :code-view, :env {:view-node 360287970189639680, :selection-provider nil, :replace-bar #object[javafx.scene.layout.GridPane 0x7edbf3ff "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :editable true, :open-resource-fn #object[clojure.core$partial$fn__5933 0x75d062d0 "clojure.core$partial$fn__5933@75d062d0"], :selection-context :code-view, :goto-line-bar #object[javafx.scene.layout.GridPane 0x47f75f0e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"], :clipboard #object[javafx.scene.input.Clipboard 0x7df480c2 "javafx.scene.input.Clipboard@7df480c2"], :prefs {:scopes {:global #object[sun.nio.fs.UnixPath 0x5ff0666b "/Users/mathiaswesterdahl/Library/Preferences/Defold/prefs.editor_settings"], :project #object[sun.nio.fs.UnixPath 0x7cb45571 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive/.editor_settings"]}, :schemas [:default #object[sun.nio.fs.UnixPath 0xa046ed5 "/Users/mathiaswesterdahl/work/projects/users/defold/extension-rive"]]}, :selection nil, :find-bar #object[javafx.scene.layout.GridPane 0x41246d4e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"]}, :selection-provider nil, :dynamics {}, :adapters {}}, :find-bar #object[javafx.scene.layout.GridPane 0x41246d4e "Grid hgap=6.0, vgap=0.0, alignment=TOP_LEFT"]}, :selection-provider nil, :dynamics {}, :adapters {}}}}
clojure.lang.ExceptionInfo: handler ':code.sort-lines' in context ':code-view' failed at ':enabled?' with message 'class clojure.lang.MapEntry cannot be cast to class editor.code.data.CursorRange (clojure.lang.MapEntry and editor.code.data.CursorRange are in unnamed module of loader 'app')'
	at editor.handler$invoke_fnk.invokeStatic(handler.clj:349)
	at editor.handler$enabled_QMARK_.invokeStatic(handler.clj:374)
	at editor.ui$make_menu_item.invokeStatic(ui.clj:1455)
	at editor.ui$make_menu_items$fn__30152.invoke(ui.clj:1472)
	at clojure.core$keep$fn__8687$fn__8688.invoke(core.clj:7517)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:418)
	at clojure.core$transduce.invokeStatic(core.clj:7025)
	at clojure.core$into.invokeStatic(core.clj:7042)
	at editor.ui$make_menu_items.invokeStatic(ui.clj:1470)
	at editor.ui$make_menu_items.invoke(ui.clj:1470)
	at editor.ui$make_menu_item.invokeStatic(ui.clj:1447)
	at editor.ui$make_menu_items$fn__30152.invoke(ui.clj:1472)
	at clojure.core$keep$fn__8687$fn__8688.invoke(core.clj:7517)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:418)
	at clojure.core$transduce.invokeStatic(core.clj:7025)
	at clojure.core$into.invokeStatic(core.clj:7042)
	at editor.ui$make_menu_items.invokeStatic(ui.clj:1470)
	at editor.ui$refresh_menubar_BANG_.invokeStatic(ui.clj:1718)
	at editor.ui$refresh_menus_BANG_.invokeStatic(ui.clj:2033)
	at editor.ui$refresh.invokeStatic(ui.clj:2066)
	at editor.app_view$refresh_menus_and_toolbars_BANG_.invokeStatic(app_view.clj:1967)
	at editor.app_view$make_app_view$fn__72930.invoke(app_view.clj:2103)
	at editor.ui$__GT_timer$fn__30535$fn__30536.invoke(ui.clj:2177)
	at clojure.lang.AFn.run(AFn.java:22)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$0(PlatformImpl.java:424)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
Caused by: java.lang.ClassCastException: class clojure.lang.MapEntry cannot be cast to class editor.code.data.CursorRange (clojure.lang.MapEntry and editor.code.data.CursorRange are in unnamed module of loader 'app')
	at editor.code.data$cursor_range_multi_line_QMARK_.invokeStatic(data.clj:208)
	at editor.code.data$cursor_range_multi_line_QMARK_.invoke(data.clj:207)
	at clojure.core$some.invokeStatic(core.clj:2718)
	at editor.code.view$can_sort_lines_QMARK_.invokeStatic(view.clj:2945)
	at editor.code.view$fn__64283$fnk64280_positional__64284.invoke(view.clj:2955)
	at editor.code.view$fn__64283$fnk64280__64295$fn__64296.invoke(view.clj:2955)
	at editor.code.view$fn__64283$fnk64280__64295.invoke(view.clj:2955)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:33)
	at clojure.lang.RestFn.invoke(RestFn.java:411)
	at editor.handler$invoke_fnk.invokeStatic(handler.clj:343)
	... 25 common frames omitted
2026-01-18 09:07:20.944 349163 [JavaFX Application Thread] INFO  editor.defold-project - {:line 375, :message "Cached loaded save data in system cache.", :total 1305, :retained 1212, :unretained 93, :limit 20000}
...
```

--- a/editor/src/clj/editor/code/resource.clj
+++ b/editor/src/clj/editor/code/resource.clj
@@ -46,7 +46,7 @@
 (g/deftype Regions [TRegion])
 (g/deftype RegionGrouping {s/Any [TRegion]})
 
-(def ^:private default-indent-type :tabs)
+(def default-indent-type :tabs)
 
 (defn make-code-error-user-data [^String path line-number]
   (let [cursor-range (some-> line-number data/line-number->CursorRange)]

--- a/editor/src/clj/editor/code/view.clj
+++ b/editor/src/clj/editor/code/view.clj
@@ -65,6 +65,7 @@
             [schema.core :as s]
             [service.smoke-log :as slog]
             [util.coll :as coll :refer [pair]]
+            [util.defonce :as defonce]
             [util.eduction :as e]
             [util.fn :as fn])
   (:import [com.defold.control ListView]
@@ -97,11 +98,11 @@
 
 (defonce ^:private default-font-size 12.0)
 
-(defprotocol GutterView
+(defonce/protocol GutterView
   (gutter-metrics [this lines regions glyph-metrics] "A two-element vector with a rounded double representing the width of the gutter and another representing the margin on each side within the gutter.")
   (draw-gutter! [this gc gutter-rect layout hovered-ui-element font color-scheme lines regions visible-cursors hovered-row] "Draws the gutter into the specified Rect."))
 
-(defrecord CursorRangeDrawInfo [type fill stroke cursor-range])
+(defonce/record CursorRangeDrawInfo [type fill stroke cursor-range])
 
 (defn- cursor-range-draw-info [type fill stroke cursor-range]
   {:pre [(case type (:range :squiggle :underline :word) true false)
@@ -163,7 +164,7 @@
             width)
           cached-width)))))
 
-(defrecord GlyphMetrics [char-width-cache ^double line-height ^double ascent]
+(defonce/record GlyphMetrics [char-width-cache ^double line-height ^double ascent]
   data/GlyphMetrics
   (ascent [_this] ascent)
   (line-height [_this] line-height)
@@ -1387,15 +1388,12 @@
   (property rename-cursor-range g/Any (dynamic visible (g/constantly false)))
   (output rename-view g/Any :cached (g/fnk [_node-id rename-cursor-range canvas-repaint-info :as props]
                                       (assoc props :fx/type rename-popup-view)))
-  (input open-views g/Any) ;; open views from app-view
 
   (output completions-selected-index g/Any :cached produce-completions-selected-index) ;; either in completions index range or nil
   (output completions-selection g/Any produce-completions-selection)
   (output completions-combined g/Any :cached produce-completions-combined)
   (output completions-combined-ids g/Any :cached produce-completions-combined-ids)
-  (input project g/NodeID) ;; used for completions doc popup, e.g. for opening defold:// URIs
 
-  (input keymap g/Any)
   (output completions-shortcut-text g/Any :cached (g/fnk [keymap]
                                                     (keymap/display-text keymap :code.show-completions nil)))
 
@@ -1411,13 +1409,19 @@
   (property visible-minimap? g/Bool (default false))
   (property visible-whitespace VisibleWhitespace (default :none))
 
-  (input completions g/Any)
-  (input cursor-ranges r/CursorRanges)
-  (input indent-type r/IndentType)
-  (input invalidated-rows r/InvalidatedRows)
+  ;; Inputs from CodeEditorResourceNode.
+  (input completions g/Any :substitute {})
+  (input cursor-ranges r/CursorRanges :substitute nil) ; Fall back on fallback-cursor-ranges.
+  (input indent-type r/IndentType :substitute r/default-indent-type)
+  (input invalidated-rows r/InvalidatedRows :substitute [])
   (input lines types/Lines :substitute [""])
-  (input regions r/Regions)
+  (input regions r/Regions :substitute [])
+
+  ;; Inputs from elsewhere.
+  (input open-views g/Any) ;; open views from app-view
+  (input project g/NodeID) ;; used for completions doc popup, e.g. for opening defold:// URIs
   (input debugger-execution-locations g/Any)
+  (input keymap g/Any)
 
   (output completion-context g/Any :cached produce-completion-context)
   (output cursor-ranges r/CursorRanges (g/fnk [cursor-ranges fallback-cursor-ranges]
@@ -3653,7 +3657,7 @@
       :ys (double-array [y1 y1 y0 y2 y4 y3 y3])
       :n  7})))
 
-(deftype CodeEditorGutterView []
+(defonce/type CodeEditorGutterView []
   GutterView
 
   (gutter-metrics [this lines regions glyph-metrics]

--- a/editor/src/dev/dev.clj
+++ b/editor/src/dev/dev.clj
@@ -306,6 +306,18 @@
   (let [wrapped-value-fn (deep-keep-kv-wrapped-value-fn value-fn)]
     (util/deep-keep-kv deep-keep-finalize-coll-value-fn wrapped-value-fn value)))
 
+(defn nodes-of-type
+  ([node-type]
+   (nodes-of-type (g/now) node-type))
+  ([basis node-type]
+   (sequence
+     (comp (map val)
+           (mapcat :nodes)
+           (map val)
+           (filter #(g/node-instance*? node-type %))
+           (map gt/node-id))
+     (:graphs basis))))
+
 (defn views-of-type [node-type]
   (keep (fn [node-id]
           (when (g/node-instance? node-type node-id)

